source/server/http/admin.cc

```diff
@@ -575,13 +574,49 @@ std::string AdminImpl::runtimeAsJson(
   return strbuf.GetString();
 }
 
+Http::Code AdminImpl::handlerHystrixEventStream(const std::string&,
+                                                Http::HeaderMap& response_headers,
+                                                Buffer::Instance&, HandlerInfo& handler_info) {
+
+  response_headers.insertContentType().value().setReference(
+      Http::Headers::get().ContentTypeValues.TextEventStream);
+  response_headers.insertCacheControl().value().setReference(
+      Http::Headers::get().CacheControlValues.NoCache);
+  response_headers.insertConnection().value().setReference(
+      Http::Headers::get().ConnectionValues.Close);
+  response_headers.insertAccessControlAllowHeaders().value().setReference(
+      Http::Headers::get().AccessControlAllowHeadersValue.AccessControlAllowHeadersHystrix);
+  response_headers.insertAccessControlAllowOrigin().value().setReference(
+      Http::Headers::get().AccessControlAllowOriginValue.All);
+  response_headers.insertNoChunks().value().setReference("0");
+
+  Stats::HystrixHandlerInfoImpl& hystrix_handler_info =
+      dynamic_cast<Stats::HystrixHandlerInfoImpl&>(handler_info);
+  Stats::HystrixHandler::HandleEventStream(&hystrix_handler_info, server_);
+
+  ENVOY_LOG(debug, "start sending data to hystrix dashboard on port {}",
+            hystrix_handler_info.callbacks_->connection()->localAddress()->asString());
+  return Http::Code::OK;
+}
+
 void AdminFilter::onComplete() {
   std::string path = request_headers_->Path()->value().c_str();
   ENVOY_STREAM_LOG(debug, "request complete: path: {}", *callbacks_, path);
 
   Buffer::OwnedImpl response;
   Http::HeaderMapPtr header_map{new Http::HeaderMapImpl};
-  Http::Code code = parent_.runCallback(path, *header_map, response);
+  Http::Code code;
+  bool end_stream = true;
+
+  if (path.find("/hystrix_event_stream") == std::string::npos) {
+    handler_info_ = std::make_unique<HandlerInfoImpl>();
+    code = parent_.runCallback(path, *header_map, response, *handler_info_);
+  } else {
+    handler_info_ = std::make_unique<Stats::HystrixHandlerInfoImpl>(callbacks_);
+    code = parent_.runCallback(path, *header_map, response, *handler_info_);
+    end_stream = false;
+  }
```

Comment:
as discussed elsewhere I am hoping to replace this whole new block with:
  Http::Code code = parent_.runCallback(path, *header_map, this, response);

Reply:
@jmarantz Thanks for the comments. Will do another fix soon.

@jmarantz @mrice32  Question for the sink implementation.
As far as I understand, currently sinks are only added to the sinks vector (MainImpl::stats_sinks_) during init, according to what's specified in the configuration file "stats_sinks" parameter.
I thought of two ways to implement:
(1) Add a "hystrix sink" through the config file. Once a request is made from the hystrix dashboard, set the connection parameters on that sink and switch on an enable flag in it.
(2) Once a request is made, create a new sink with the connection parameters and add it to the vector of sinks. Remove it from the vector when the connection is destroyed.
Do you see any problem or disadvantage with any of these options?

Reply:
@trabetti, 1 is almost surely what we want IIUC.  I'm not exactly sure what you mean by setting connection parameters or enabling the flag, but by my understanding, you should be able to just register each stream with the single hystrix sink that is created through the config.  We'll also need to detect connection termination and make sure the sink is aware of this so it doesn't try to access a deleted stream. This registration mechanism will take some thought, by the way, because there is no existing communication mechanism between stats sinks and the admin interface.

As you're working through this design change, feel free to reach out with any questions here or on slack.  We're happy to provide input wherever needed :).  Take your time making incremental updates, here, and whenever you're ready for us to take another pass, just let us know.

Reply:
Hi @mrice32 , the first version of the hystrix support as a sink is almost ready, I hope I'll get to push it before Passover holiday..
Questions:
1. As we discussed, the hystrix sink will be set through the config file, which means I touched data_plane_api. How do I push it here so it compiles (currently I point to data_plane_api branch in my local fork)
2. I need some help figuring out how to use the histogram data. 
Thanks!

Reply:
1.  You should send out a PR for your data-plane-api changes, which will need to be merged before this PR is merged.  Until that PR gets merged, you should do exactly what you're doing now - point the tests to your own fork to allow the CI to pass in the meantime.

2.  What, specifically, about the histogram data is confusing?  The model for histograms is different for the other stats.  Each sample is sent to the sinks when published.  This is done synchronously by calling [this method](https://github.com/envoyproxy/envoy/blob/master/include/envoy/stats/stats.h#L206) from whatever thread recorded the value.  So you need to be sure that that method is thread-safe.  The statsd implementation uses TLS to ensure thread-safety (see [here](https://github.com/envoyproxy/envoy/blob/master/source/common/stats/statsd.cc#L64)), but since you are probably interested in forming some sort of central cache from which to pull at flush time, your implementation will likely be more complicated than that.  It's also worth noting that there's [work underway](https://github.com/envoyproxy/envoy/issues/1965#issuecomment-376672282) on a design to create an Envoy representation of the distributions and central cache to store them as opposed to just handing the samples to the sinks as is done now.